File: src/turtlebot/voice_teleop/voice_teleop.cpp
/*
 *  Gaitech Educational Portal
 *
 * Copyright (c) 2016
 * All rights reserved.
 *
 * License Type: GNU GPL
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include <ros/ros.h>
#include <geometry_msgs/Twist.h>
#include <std_msgs/String.h>
#include "boost/thread/mutex.hpp"
#include "boost/thread/thread.hpp"
#include "ros/console.h"
#include <string>

std_msgs::String readCommand;

class RobotVoiceTeleop
{
public:
	RobotVoiceTeleop();

private:
	void commandCallBack(const std_msgs::String& command);
	void publish();

	ros::NodeHandle nodeHandle;


	int linearVelocity, angularVelocity;
	ros::Publisher velocityPublisher;
	ros::Subscriber voiceCommandSubscriber;
	geometry_msgs::Twist cmd_vel;


};
/** constructor **/
RobotVoiceTeleop::RobotVoiceTeleop():
						  nodeHandle("~"),
						  linearVelocity(1),
						  angularVelocity(0)
{
	nodeHandle.param("linear_velocity", linearVelocity, linearVelocity);
	nodeHandle.param("angular_velocity", angularVelocity, angularVelocity);

	velocityPublisher = nodeHandle.advertise<geometry_msgs::Twist>("/cmd_vel", 1);
	voiceCommandSubscriber = nodeHandle.subscribe("/recognizer/output", 1000, &RobotVoiceTeleop::commandCallBack, this);

	printf("start control\n");
	ros::Rate rate(5);
	while (ros::ok()){
		velocityPublisher.publish(cmd_vel);
		rate.sleep();
		ros::spinOnce();
	}

}
/** callback **/
void RobotVoiceTeleop::commandCallBack(const std_msgs::String& command)
{ 
	if(command.data.compare("stop") == 0)
	{
		cmd_vel.angular.z = 0;
		cmd_vel.linear.x = 0;
	}
	else if(command.data.compare("forward") == 0)
	{
		cmd_vel.linear.x = 0.3;
		cmd_vel.angular.z = 0;
	}
	else if(command.data.compare("backward") == 0)
	{
		cmd_vel.linear.x = -0.3;
		cmd_vel.angular.z = 0;
	}
	else if(command.data.compare("left") == 0)
	{
		cmd_vel.linear.x = 0.0;
		cmd_vel.angular.z = 0.5;
	}
	else if(command.data.compare("right") == 0)
	{
		cmd_vel.linear.x = 0.0;
		cmd_vel.angular.z = -0.5;
	}
	else {
		cmd_vel.angular.z = 0;
		cmd_vel.linear.x = 0;
	}

	printf("cmd_vel.linear.x = %.2f, cmd_vel.angular.z = %.2f\n", cmd_vel.linear.x, cmd_vel.angular.z);

}


int main(int argc, char** argv)
{
	ros::init(argc, argv, "voice_teleop_node");
	RobotVoiceTeleop voice_teleop;
	ros::spin();
}

File: src/turtlebot/voice_teleop/voice_teleop.py
#!/usr/bin/env python

import rospy
from geometry_msgs.msg import Twist
from std_msgs.msg import String

class RobotVoiceTeleop:
    #define the constructor of the class
    def  __init__(self):
        #initialize the ROS node with a name voice_teleop
        rospy.init_node('voice_teleop')
        
        # Publish the Twist message to the cmd_vel topic
        self.cmd_vel_pub = rospy.Publisher('/cmd_vel', Twist, queue_size=5)
        
        # Subscribe to the /recognizer/output topic to receive voice commands.
        rospy.Subscriber('/recognizer/output', String, self.voice_command_callback)
        
        #create a Rate object to sleep the process at 5 Hz
        rate = rospy.Rate(5)
        
        # Initialize the Twist message we will publish.
        self.cmd_vel = Twist()
        #make sure to make the robot stop by default
        self.cmd_vel.linear.x=0;
        self.cmd_vel.angular.z=0;
        
        
        
        # A mapping from keywords or phrases to commands
        #we consider the following simple commands, which you can extend on your own
        self.commands =             ['stop',
                                    'forward',
                                    'backward',
                                    'turn left',
                                    'turn right',
                                    ]
        rospy.loginfo("Ready to receive voice commands")
        # We have to keep publishing the cmd_vel message if we want the robot to keep moving.
        while not rospy.is_shutdown():
            self.cmd_vel_pub.publish(self.cmd_vel)
            rate.sleep()


    def voice_command_callback(self, msg):
        # Get the motion command from the recognized phrase
        command = msg.data
        if (command in self.commands):
            if command == 'forward':
                self.cmd_vel.linear.x = 0.2
                self.cmd_vel.angular.z = 0.0
            elif command == 'backward':
                self.cmd_vel.linear.x = -0.2
                self.cmd_vel.angular.z = 0.0
            elif command == 'turn left':
                self.cmd_vel.linear.x = 0.0
                self.cmd_vel.angular.z = 0.5
            elif command == 'turn right':
                self.cmd_vel.linear.x = 0.0
                self.cmd_vel.angular.z = -0.5
            elif command == 'stop':
                self.cmd_vel.linear.x = 0.0
                self.cmd_vel.angular.z = 0.0

        else: #command not found
            #print 'command not found: '+command
            self.cmd_vel.linear.x = 0.0
            self.cmd_vel.angular.z = 0.0
        print ("linear speed : " + str(self.cmd_vel.linear.x))
        print ("angular speed: " + str(self.cmd_vel.angular.z))



if __name__=="__main__":
    try:
      RobotVoiceTeleop()
      rospy.spin()
    except rospy.ROSInterruptException:
      rospy.loginfo("Voice navigation terminated.")

